src/utils/intrusive_ptr_test.cc

```diff
@@ -0,0 +1,155 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#include "src/utils/intrusive_ptr.h"
+
+#include <iostream>
+#include <string>
+
+#include "src/common/testing/gtest.h"
+
+namespace raksha {
+namespace {
+
+class RefCountedType {
+ public:
+  RefCountedType(std::string id = "<unknown>", unsigned int initial_count = 0)
+      : id_(id), count_(initial_count) {}
+
+  unsigned int count() const { return count_; }
+  void set_id(std::string id) { std::swap(id_, id); }
+  const std::string& id() const { return id_; }
+  void Retain() { count_++; }
+  void Release() { count_--; }
+
+ private:
+  std::string id_;
+  unsigned int count_;
+};
+
+using RefCountedTypePtr = intrusive_ptr<RefCountedType>;
+
+TEST(IntrusivePtrTest, DefaultConstructorInitializesToNullptr) {
+  RefCountedTypePtr null;
+  EXPECT_EQ(null.get(), nullptr);
+}
+
+TEST(IntrusivePtrTest, ConstructorIncreasesCountAndSavesPtr) {
+  auto value = std::make_unique<RefCountedType>();
+  EXPECT_EQ(value->count(), 0);
+
+  RefCountedTypePtr ptr(value.get());
+  EXPECT_EQ(value->count(), 1);
+
+  EXPECT_EQ(ptr.get(), value.get());
+}
+
+TEST(IntrusivePtrTest, CopyConstructorIncreasesCountAndSharesPtr) {
+  auto value = std::make_unique<RefCountedType>();
+  EXPECT_EQ(value->count(), 0);
+
+  RefCountedTypePtr ptr(value.get());
+  EXPECT_EQ(value->count(), 1);
+
+  RefCountedTypePtr copy = ptr;
+  EXPECT_EQ(value->count(), 2);
+  EXPECT_EQ(ptr.get(), copy.get());
+}
+
+TEST(IntrusivePtrTest, MoveConstructorMaintainsCountAndNullifiesOther) {
+  auto value = std::make_unique<RefCountedType>();
+  EXPECT_EQ(value->count(), 0);
+
+  RefCountedTypePtr ptr(value.get());
+  EXPECT_EQ(value->count(), 1);
+
+  RefCountedTypePtr moved = std::move(ptr);
+  EXPECT_EQ(value->count(), 1);
+  EXPECT_EQ(ptr.get(), nullptr);
+}
```

Comment:
This pattern you use in this file of having a `unique_ptr` and an intrusive pointer point to the same value is very easy to misunderstand. I think it could use some comments documenting its behavior.

Reply:
Good point. It is not good idea to leave bad code examples in the test. I refactored this without `unique_ptr` and documented why we are using raw pointer for tests.